extmod/nimble/modbluetooth_nimble.c

```diff
@@ -1110,4 +1115,281 @@ int mp_bluetooth_gattc_exchange_mtu(uint16_t conn_handle) {
+                os_mbuf_concat(chan->rx_pending, event->receive.sdu_rx);
+            } else {
+                // Normal case is when the first payload arrives since calling
+                // ble_l2cap_recv_ready.
+                DEBUG_printf("l2cap_event: receive: new payload\n");
+                // Take ownership of sdu_rx.
+                chan->rx_pending = event->receive.sdu_rx;
+            }
+
+            struct os_mbuf *sdu_rx = os_mbuf_get_pkthdr(&chan->sdu_mbuf_pool, 0);
+            assert(sdu_rx);
+            if (!sdu_rx) {
+                printf("data received: no sdu_rx\n");
+            }
+
+            // ble_l2cap_coc_rx_fn invokes this event handler when a complete payload arrives.
+            // However, it NULLs chan->chan->coc_rx.sdu before doing so, expecting that
+            // ble_l2cap_recv_ready will be called to give it a new mbuf.
+            // This means that if another payload arrives before we call ble_l2cap_recv_ready
+            // then ble_l2cap_coc_rx_fn will NULL-deref coc_rx.sdu.
+
+            // Because we're not yet ready to grant new credits to the channel, we can't call
+            // ble_l2cap_recv_ready yet, so instead we just give it a new mbuf. This requires
+            // ble_l2cap_priv.h for the definition of chan->chan (i.e. struct ble_l2cap_chan).
+            chan->chan->coc_rx.sdu = sdu_rx;
+
+            mp_bluetooth_gattc_on_l2cap_coc_generic(MP_BLUETOOTH_IRQ_L2CAP_COC_RECV, L2CAP_COC_CHAN_ID);
+            break;
+        }
+        case BLE_L2CAP_EVENT_COC_TX_UNSTALLED: {
+            DEBUG_printf("l2cap_coc_event: tx_unstalled: conn_handle=%d status=%d\n", event->tx_unstalled.conn_handle, event->tx_unstalled.status);
+            mp_bluetooth_gattc_on_l2cap_coc_tx_unstalled(L2CAP_COC_CHAN_ID, event->tx_unstalled.status);
+            break;
+        }
+        case BLE_L2CAP_EVENT_COC_RECONFIG_COMPLETED: {
+            DEBUG_printf("l2cap_coc_event: reconfig_completed: conn_handle=%d\n", event->reconfigured.conn_handle);
+            break;
+        }
+        case BLE_L2CAP_EVENT_COC_PEER_RECONFIGURED: {
+            DEBUG_printf("l2cap_coc_event: peer_reconfigured: conn_handle=%d\n", event->reconfigured.conn_handle);
+            break;
+        }
+        default: {
+            DEBUG_printf("l2cap_coc_event: unknown event\n");
+            break;
+        }
+    }
+
+    return 0;
+}
+
+STATIC mp_bluetooth_nimble_l2cap_coc_t *get_l2cap_coc_from_handle(uint16_t chan_handle) {
+    if (chan_handle != L2CAP_COC_CHAN_ID) {
+        // TODO: Support more than one concurrent L2CAP channel.
+        return NULL;
+    }
+    return MP_STATE_PORT(bluetooth_nimble_root_pointers)->l2cap_chan;
+}
+
+STATIC int create_l2cap_coc(uint16_t mtu, mp_bluetooth_nimble_l2cap_coc_t **out) {
+    if (MP_STATE_PORT(bluetooth_nimble_root_pointers)->l2cap_chan) {
+        // Only one L2CAP channel allowed.
+        // Additionally, if we're listening, then no connections may be initiated.
```

Comment:
I guess this is just a limitation of this implementation, not of BLE L2CAP in general.  And we can eventually support multiple L2CAP connections at once?

Reply:
That's right. We'll need mp_bluetooth_nimble_root_pointers_t to store a map of `(conn_handle,cid)` -> `channel` instances.